crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs

```diff
@@ -74,6 +74,14 @@ pub enum SenderData {
         /// If false, we had simply accepted the key as this user's latest
         /// key.
         master_key_verified: bool,
+
+        #[serde(default)]
+        /// Whether the user's master key was previously verified, but the
+        /// current master key is not.
+        ///
+        /// TODO: should this be merged with the master_key_verified field and
+        /// turned into an enum?
+        previously_verified: bool,
```

Comment:
hrm, this feels wrong. 

"Previously-verified-but-no-longer" is surely a property of the sender, not the session? Not least because all received messages are going to get stuck in that state even once we verify the new identity.

But perhaps that's behaviour we want?

Reply:
We need to store `previously_verified` (and, for that matter, `master_key_verified`, and the whole of `SenderKnown`-ness) in SenderData rather than in the user's identity because the session data needs to latch towards is most-trusted state. 

Imagine we see *3* different identities for Alice. Initially she has identity A, which is verified. We then see identity B; while that is active, she sends a message. Finally, we see a third identity, C.

For that message sent under identity B, we must remember whether the identity was verified or not, as well as the fact that it followed an identity A that was previously verified.

We cannot simply rely on the verification properties of the *current* identity C.

Reply:
Next question: is it really a property of a SenderKnown session? Arguably the sender isn't really known, because it's a bad state; so we should have a separate enum variant for this case. Also, it would keep clean the interface to CryptoStore for the update logic on /keys/query (
https://github.com/matrix-org/matrix-rust-sdk/issues/3753#issuecomment-2263389209).

So we need a new enum variant called something like `SenderKnownButUnverified`.

Question: should it be `SenderKnownButUnverified` or `SenderKnownButUnverifiedAndPreviouslyVerified` (ie, where do sessions from tofu-trusted identities end up, and which of the variants has a `master_key_verified` field)

Arguably some of the logic is easier with `SenderKnownButUnverified` - and we get a bonus of updating tofu-trusted sessions to verified sessions. But it conflates known-bad state (no longer verified) with ok state (tofu-verified). Also updating all the tofu-trusted sessions could be prohibitively expensive.

What if we had *five* variants? It leaves the update algorithm open for later choice.

... We agree we want five variants. @uhoreg will make a separate PR to create new variants.

Reply:
> We need to store `previously_verified` (and, for that matter, `master_key_verified`, and the whole of `SenderKnown`-ness) in SenderData rather than in the user's identity because the session data needs to latch towards is most-trusted state.
> 
> Imagine we see _3_ different identities for Alice. Initially she has identity A, which is verified. We then see identity B; while that is active, she sends a message. Finally, we see a third identity, C.
> 
> For that message sent under identity B, we must remember whether the identity was verified or not, as well as the fact that it followed an identity A that was previously verified.
> 
> We cannot simply rely on the verification properties of the _current_ identity C.

It would be good to try to distill this train-of-thought into a useful comment in the code.